File: 工厂方法模式.cpp
#include <bits/stdc++.h>
using namespace std;
class AbstractFruit
{
public:
    virtual void ShowName() = 0;
};
class Apple : public AbstractFruit
{
public:
    virtual void ShowName()
    {
        cout << "I'm apple!" << endl;
    }
};
class Banana : public AbstractFruit
{
public:
    virtual void ShowName()
    {
        cout << "I'm banana!" << endl;
    }
};
class AbstractFactory
{
public:
    virtual AbstractFruit *Createfruit() = 0;
};
class AppleFactory : public AbstractFactory
{
public:
    virtual AbstractFruit *Createfruit()
    {
        return new Apple;
    }
};
class BananaFactory : public AbstractFactory
{
public:
    virtual AbstractFruit *Createfruit()
    {
        return new Banana;
    }
};
void test()
{
    AbstractFactory *factory = new AppleFactory;
    AbstractFruit *fruit = factory->Createfruit();
    fruit->ShowName();
    delete fruit;
    delete factory;
    fruit = nullptr;
    factory = nullptr;
}
int main()
{
    test();
    return 0;
}

File: 工厂方法模式.go
//go:build ignore

package main

import "fmt"

type Abstractfruit interface {
	ShowName()
}
type Apple struct {
}

func (apple *Apple) ShowName() {
	fmt.Println("I'm apple")
}

type Banana struct {
}

func (banana *Banana) ShowName() {
	fmt.Println("I'm banana")
}

type Abstractfactory interface {
	Createfruit() Abstractfruit
}
type Applefactory struct {
}

func (applefactory *Applefactory) Createfruit() Abstractfruit {
	return new(Apple)
}

type Bananafactory struct {
}

func (bananafactory *Bananafactory) Createfruit() Abstractfruit {
	return new(Banana)
}
func main() {
	var factory Abstractfactory
	var fruit Abstractfruit
	factory = new(Applefactory)
	fruit = factory.Createfruit()
	fruit.ShowName()
	factory = new(Bananafactory)
	fruit = factory.Createfruit()
	fruit.ShowName()
}
